File: Advanced/duediligence-cordapp/workflows/src/main/java/net/corda/samples/duediligence/flows/AddApprovalsToCheckList.java
package net.corda.samples.duediligence.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.flows.*;
import net.corda.core.identity.Party;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.samples.duediligence.contracts.DueDChecklistContract;
import net.corda.samples.duediligence.states.CopyOfCoporateRecordsAuditRequest;
import net.corda.samples.duediligence.states.CorporateRecordsAuditRequest;
import net.corda.samples.duediligence.states.DueDChecklist;

import javax.naming.ldap.UnsolicitedNotification;
import java.util.Arrays;
import java.util.UUID;

public class AddApprovalsToCheckList {

    @InitiatingFlow
    @StartableByRPC
    public static class CreateCheckListAndAddApprovalInitiator extends FlowLogic<SignedTransaction> {

        private Party reportTo;
        private UniqueIdentifier approvalId;

        public CreateCheckListAndAddApprovalInitiator(Party reportTo, UniqueIdentifier approvalId) {
            this.reportTo = reportTo;
            this.approvalId = approvalId;
        }

        @Override
        @Suspendable
        public SignedTransaction call() throws FlowException {

            //Query the input
            QueryCriteria.LinearStateQueryCriteria inputCriteria = new QueryCriteria.LinearStateQueryCriteria()
                    .withUuid(Arrays.asList(UUID.fromString(approvalId.toString())))
                    .withStatus(Vault.StateStatus.UNCONSUMED)
                    .withRelevancyStatus(Vault.RelevancyStatus.RELEVANT);
            StateAndRef inputStateAndRef = getServiceHub().getVaultService().queryBy(ContractState.class, inputCriteria).getStates().get(0);

            //extract the notary
            Party notary = inputStateAndRef.getState().getNotary();


            //create due-diligence Checklist
            DueDChecklist checklist = new DueDChecklist(3,getOurIdentity(),reportTo,new UniqueIdentifier());
            checklist.uploadApproval(this.approvalId);

            TransactionBuilder txBuilder = new TransactionBuilder(notary)
                    .addInputState(inputStateAndRef)
                    .addOutputState(checklist)
                    .addCommand(new DueDChecklistContract.Commands.Add(),Arrays.asList(getOurIdentity().getOwningKey(),reportTo.getOwningKey()));

            // Verify that the transaction is valid.
            txBuilder.verify(getServiceHub());

            // Sign the transaction.
            final SignedTransaction partSignedTx = getServiceHub().signInitialTransaction(txBuilder);

            // Send the state to the counterparty, and receive it back with their signature.
            FlowSession otherPartySession = initiateFlow(reportTo);
            final SignedTransaction fullySignedTx = subFlow(
                    new CollectSignaturesFlow(partSignedTx, Arrays.asList(otherPartySession), CollectSignaturesFlow.Companion.tracker()));

            // Notarise and record the transaction in both parties' vaults.
            if (inputStateAndRef.getState().getData().getClass() == CorporateRecordsAuditRequest.class){
                CorporateRecordsAuditRequest request = (CorporateRecordsAuditRequest) inputStateAndRef.getState().getData();
                return subFlow(new FinalityFlow(fullySignedTx, Arrays.asList(otherPartySession,initiateFlow(request.getValidater()))));
            }else{
                CopyOfCoporateRecordsAuditRequest request = (CopyOfCoporateRecordsAuditRequest) inputStateAndRef.getState().getData();
                return subFlow(new FinalityFlow(fullySignedTx, Arrays.asList(otherPartySession,initiateFlow(request.getoriginalValidater()),initiateFlow(request.getOriginalOwner()))));
            }
        }
    }

    @InitiatedBy(CreateCheckListAndAddApprovalInitiator.class)
    public static class CreateCheckListAndAddApprovalResponder extends FlowLogic<Void> {
        //private variable
        private FlowSession counterpartySession;

        //Constructor
        public CreateCheckListAndAddApprovalResponder(FlowSession counterpartySession) {
            this.counterpartySession = counterpartySession;
        }

        @Suspendable
        @Override
        public Void call() throws FlowException {
            SignedTransaction signedTransaction = subFlow(new SignTransactionFlow(counterpartySession) {
                @Suspendable
                @Override
                protected void checkTransaction(SignedTransaction stx) throws FlowException {
                }
            });
            //Stored the transaction into data base.
            subFlow(new ReceiveFinalityFlow(counterpartySession, signedTransaction.getId()));
            return null;
        }
    }

}

File: kotlin-samples/Advanced/duediligence-cordapp/workflows/src/main/kotlin/net/corda/samples/duediligence/flows/AddApprovalsToCheckList.kt
package net.corda.samples.duediligence.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.contracts.ContractState
import net.corda.core.contracts.StateAndRef
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.flows.*
import net.corda.core.flows.CollectSignaturesFlow.Companion.tracker
import net.corda.core.identity.Party
import net.corda.core.node.services.Vault.RelevancyStatus
import net.corda.core.node.services.Vault.StateStatus
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.samples.duediligence.contracts.DueDChecklistContract
import net.corda.samples.duediligence.states.CopyOfCoporateRecordsAuditRequest
import net.corda.samples.duediligence.states.CorporateRecordsAuditRequest
import net.corda.samples.duediligence.states.DueDChecklist
import java.security.PublicKey
import java.util.*

@InitiatingFlow
@StartableByRPC
class CreateCheckListAndAddApprovalInitiator(private val reportTo: Party, private val approvalId: UniqueIdentifier) : FlowLogic<SignedTransaction?>() {
    @Suspendable
    @Throws(FlowException::class)
    override fun call(): SignedTransaction {

        //Query the input
        val inputCriteria: QueryCriteria.LinearStateQueryCriteria = QueryCriteria.LinearStateQueryCriteria()
                .withUuid(Arrays.asList(UUID.fromString(approvalId.toString())))
                .withStatus(StateStatus.UNCONSUMED)
                .withRelevancyStatus(RelevancyStatus.RELEVANT)
        val inputStateAndRef: StateAndRef<*> = serviceHub.vaultService.queryBy<ContractState>(ContractState::class.java, inputCriteria).states.get(0)

        //extract the notary
        val notary = inputStateAndRef.state.notary


        //create due-diligence Checklist
        val checklist = DueDChecklist(numberOfapprovalsNeeded = 3,
                operationNode = ourIdentity,
                reportTo = reportTo,
                linearId = UniqueIdentifier())
        checklist.uploadApproval(approvalId)
        val txBuilder: TransactionBuilder = TransactionBuilder(notary)
                .addInputState(inputStateAndRef)
                .addOutputState(checklist)
                .addCommand(DueDChecklistContract.Commands.Add(), Arrays.asList(ourIdentity.owningKey, reportTo.owningKey))

        // Verify that the transaction is valid.
        txBuilder.verify(serviceHub)

        // Sign the transaction.
        val partSignedTx = serviceHub.signInitialTransaction(txBuilder)

        // Send the state to the counterparty, and receive it back with their signature.
        val otherPartySession = initiateFlow(reportTo)
        val fullySignedTx = subFlow(
                CollectSignaturesFlow(partSignedTx, Arrays.asList(otherPartySession), tracker()))

        // Notarise and record the transaction in both parties' vaults.
        return if (inputStateAndRef.state.data.javaClass == CorporateRecordsAuditRequest::class.java) {
            val request = inputStateAndRef.state.data as CorporateRecordsAuditRequest
            subFlow(FinalityFlow(fullySignedTx, Arrays.asList(otherPartySession, initiateFlow(request.validater))))
        } else {
            val request = inputStateAndRef.state.data as CopyOfCoporateRecordsAuditRequest
            subFlow(FinalityFlow(fullySignedTx, Arrays.asList(otherPartySession, initiateFlow(request.originalValidater), initiateFlow(request.originalOwner))))
        }
    }
}

@InitiatedBy(CreateCheckListAndAddApprovalInitiator::class)
class CreateCheckListAndAddApprovalResponder
(
        private val counterpartySession: FlowSession) : FlowLogic<SignedTransaction>() {
    @Suspendable
    @Throws(FlowException::class)
    override fun call(): SignedTransaction {
        val signedTransaction = subFlow(object : SignTransactionFlow(counterpartySession) {
            @Suspendable
            @Throws(FlowException::class)
            override fun checkTransaction(stx: SignedTransaction) {
            }
        })
        //Stored the transaction into data base.
        return subFlow(ReceiveFinalityFlow(counterpartySession, signedTransaction.id))
    }
}
